src/wp-admin/includes/class-wp-upgrader.php

```diff
@@ -811,6 +844,20 @@ public function run( $options ) {
 
 		$this->skin->set_result( $result );
 		if ( is_wp_error( $result ) ) {
+			if ( ! empty( $options['hook_extra']['temp_backup'] ) ) {
+				/*
+				 * Restore the backup on shutdown.
+				 * Actions running on `shutdown` are immune to PHP timeouts,
+				 * so in case the failure was due to a PHP timeout,
+				 * we'll still be able to properly restore the previous version.
+				 */
+				add_action(
+					'shutdown',
+					function() use ( $options ) {
+						$this->restore_temp_backup( $options['hook_extra']['temp_backup'] );
+					}
+				);
```

Comment:
2 issues here:

- Avoid the word "we" in the comment as it's unclear who "we" is referring to.

- A closure as a filter/action callback is not allowed [per coding standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#closures-anonymous-functions).
>Closures must not be passed as filter or action callbacks, as they cannot be removed by remove_action() / remove_filter() (see [#46635](https://core.trac.wordpress.org/ticket/46635) for a proposal to address this).

Instead, add a private method and store the `$options` as a private property, i.e. so that it's available to the showdown method.

Reply:
I thought closures were allowed if they were just calling functions and not modifying data.

Reply:
Closures removed.